### PR TITLE
feat(`signer`): sign dynamic typed data

### PR DESCRIPTION
**Pull request by youngkashew**
## Motivation
Resolves issue #202.

## Solution
Add methods `sign_dynamic_typed_data` and `sign_dynamic_typed_data_sync` to the `Signer` and `SignerSync` traits, respectively.

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -41,7 +41,7 @@ alloy-transport-ipc = { version = "0.1.0", default-features = false, path = "cra
 alloy-transport-ws = { version = "0.1.0", default-features = false, path = "crates/transport-ws" }
 
 alloy-core = { version = "0.6.3", default-features = false, features = ["std"] }
-alloy-dyn-abi = { version = "0.6.3", default-features = false, features = ["std"] }
+alloy-dyn-abi = { version = "0.6.3", default-features = false, features = ["std", "eip712"] }
```
Comment (prestwich) [inline Cargo.toml:44]: can remove this and address it in the signers

File: crates/signer/Cargo.toml [hunk 2]
```diff
@@ -24,6 +24,7 @@ async-trait.workspace = true
 
 # eip712
 alloy-sol-types = { workspace = true, optional = true }
```
Comment (prestwich) [inline crates/signer/Cargo.toml:26]: feature requirement goes here

File: Cargo.toml (hunk 1 above)
Comment (youngkashew) [reply Cargo.toml]: Resolved, thanks.

File: crates/signer/Cargo.toml (hunk 2 above)
Comment (youngkashew) [reply crates/signer/Cargo.toml:26]: Resolved, thanks.

File: crates/signer/src/signer.rs [hunk 3]
```diff
@@ -100,6 +101,17 @@ pub trait Signer: Send + Sync {
...
     }
 
+    /// Encodes and signs the typed data according to [EIP-712] for Signers that are not dynamically sized.
+    #[cfg(feature = "eip712")]
+    #[inline]
+    async fn sign_dynamic_typed_data(
+        &self,
+        payload: &TypedData
+    ) -> Result<Signature> {
+        let hash = payload.eip712_signing_hash().unwrap();
```
Comment (DaniPopes) [inline crates/signer/src/signer.rs:111]: Error should be propagated, not unwrapped. You need to add a variant to the `Error` enum https://github.com/alloy-rs/alloy/blob/7b35f83099f78dcebfb5d1075f2245a85380a774/crates/signer/src/error.rs#L11

File: crates/signer/src/signer.rs [hunk 4]
```diff
@@ -183,6 +195,19 @@ pub trait SignerSync {
...
+    /// Encodes and signs the typed data according to [EIP-712] for Signers that are not dynamically sized.
+    ///
+    /// [EIP-712]: https://eips.ethereum.org/EIPS/eip-712
+    #[cfg(feature = "eip712")]
+    #[inline]
+    fn sign_dynamic_typed_data_sync(
+        &self,
+        payload: &TypedData
+    ) -> Result<Signature> {
+        let hash = payload.eip712_signing_hash().unwrap();
```
Comment (DaniPopes) [inline crates/signer/src/signer.rs:207]: Same here

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/signer/src/signer.rs (hunk 3 above)
Comment (youngkashew) [reply crates/signer/src/signer.rs]: Resolved, thank you

File: crates/signer/src/signer.rs (hunk 4 above)
Comment (youngkashew) [reply crates/signer/src/signer.rs]: Resolved, thank you

**Comment by youngkashew** [comment]
Fixed issues with the formatting. It looks like the CI tests are failing because of a compilation issue:
```
error[E0432]: unresolved import `alloy_sol_types`
 --> crates/signer/src/signer.rs:8:5
  |
8 | use alloy_sol_types::{Eip712Domain, SolStruct};
  |     ^^^^^^^^^^^^^^^ use of undeclared crate or module `alloy_sol_types`
```
I don't believe I made any changes that caused this. Anything I need to should be doing to fix this @DaniPopes?

File: crates/signer/src/signer.rs [hunk 5]
```diff
@@ -4,6 +4,7 @@ use async_trait::async_trait;
 use auto_impl::auto_impl;
 
 #[cfg(feature = "eip712")]
+use alloy_dyn_abi::eip712::TypedData;
 use alloy_sol_types::{Eip712Domain, SolStruct};
```
Comment (DaniPopes) [inline crates/signer/src/signer.rs:8]: This is missing cfg

**Review by DaniPopes — APPROVED** [review]
Thanks! Missing just formatting.

**Comment by youngkashew** [comment]
> Thanks! Missing just formatting.

@DaniPopes Fixed! Thanks for your help.

**Review by onbjerg — APPROVED** [review]
needs a rebase (sorry about that), otherwise lgtm

**Comment by youngkashew** [comment]
> needs a rebase (sorry about that), otherwise lgtm

@onbjerg Thanks for the review. I believe I resolved all of the merge conflicts.

**Review by onbjerg — APPROVED** [review]

**Pull request merged**